### PR TITLE
dev/financial#150: civicrm/payment/form url got empty currency argument in backoffice live CC form

### PR DESCRIPTION
**Pull request by monishdeb**
Overview
----------------------------------------
This affects iATS Payments ACH/EFT payment processor which renders the payment block based on currency chosen. 
Here are the steps to replicate:

1. Install and enable iATS paymentextension
2. Add and configure an iATS Payments ACH/EFT payment type payment processor.
3. Enable USD and CAD currencies
4. Go to 'Submit Credit Card Contribution' backoffice form
5. Choose payment processor created at step 2

Issue:
1. Payment block renders payment block without check template
2. Switching currency doesn't update the payment block.

Before
----------------------------------------
crm-598-before

After
----------------------------------------
crm-598-after

Technical Details
----------------------------------------
For more detail discussion on MM - https://chat.civicrm.org/civicrm/pl/ixxrxhxs43r4mfyjpnf6tnwnnc

https://lab.civicrm.org/dev/financial/-/issues/150

ping @KarinG @Edzelopez @eileenmcnaughton @JoeMurray

### CODE REVIEW

File: CRM/Contribute/Form/AbstractEditPayment.php [hunk 1]
```diff
@@ -385,19 +385,7 @@ public function assignProcessors() {
...
-      $config->defaultCurrency
-    );
-
-    // use submitted currency if present else use current currency
-    $result = CRM_Utils_Array::value('currency',
-      $submittedValues,
-      $currentCurrency
-    );
-    return $result;
+    return $submittedValues['currency'] ?? $this->_values['currency'] ?? CRM_Core_Config::singleton()->defaultCurrency;
```
Comment (mattwire) [inline CRM/Contribute/Form/AbstractEditPayment.php:388]: This part looks good - it's just cleanup.

File: templates/CRM/common/paymentBlock.tpl [hunk 2]
```diff
@@ -95,7 +95,10 @@
 
       var payment_instrument_id = $('#payment_instrument_id').val();
 
-      var dataUrl = "{crmURL p='civicrm/payment/form' h=0 q="formName=`$form.formName`&currency=`$currency`&`$urlPathVar``$isBackOfficePathVar``$profilePathVar``$contributionPageID``$preProfileID`processor_id="}" + type;
+      var currency = '{$currency}';
+      currency = currency == '' ? $('#currency').val() : currency;
```
Comment (mattwire) [inline templates/CRM/common/paymentBlock.tpl:99]: This is the only bit I have concerns with - why would the currency not be set when the form loads?

**Comment by mattwire** [comment]
I've raised a couple of comments for an `r-code` review. I haven't read through the detail behind this PR so they may have already been answered.

File: templates/CRM/common/paymentBlock.tpl (hunk 2 above)
Comment (monishdeb) [reply templates/CRM/common/paymentBlock.tpl:99]: @mattwire actually there are two cases where the currency variable is not set in backend, these are:
 
1. When we open the standalone CC form for submitting new live contribution, where we are setting a NULL currency value in [here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/Form/Contribution.php#L429) Although the default currency set in the system can be used in here instead if NULL. So I am not sure if we need to do that.

2. When we are changing the  currency beside the amount field https://github.com/civicrm/civicrm-core/pull/18517/files#diff-726c3dcc8c2364a0b9b8f0c0e11bcd34R122 which call the js function, and in this case the code should consider what currency value has been chosen in b/o form and when in reloads the snipper 

This is not a issue on edit CC form where currency variable is set.

**Comment by eileenmcnaughton** [comment]
test this please

File: templates/CRM/common/paymentBlock.tpl (hunk 2 above)
Comment (eileenmcnaughton) [reply templates/CRM/common/paymentBlock.tpl:99]: @monishdeb can we pass currency though the url for case 2?

**Comment by KarinG** [comment]
When Jenkins is happy I'll test it.

**Comment by eileenmcnaughton** [comment]
thanks @KarinG

File: templates/CRM/common/paymentBlock.tpl [hunk 3]
```diff
@@ -108,15 +111,15 @@
...
 
       // Processors like pp-express will hide the form submit buttons, so re-show them when switching
       $('.crm-submit-buttons', $form).show().find('input').prop('disabled', true);
       CRM.loadPage(dataUrl, {target: '#billing-payment-block'});
     }
 
-    $('[name=payment_processor_id]').on('change.paymentBlock', function() {
-        buildPaymentBlock($(this).val());
+    $('[name=payment_processor_id], #currency').on('change.paymentBlock', function() {
+      buildPaymentBlock($('[name=payment_processor_id]').val());
```
Comment (samuelsov) [inline templates/CRM/common/paymentBlock.tpl:122]: We need to add `:checked` or it will always take the first option.

Comment (samuelsov) [reply templates/CRM/common/paymentBlock.tpl:122]: Tested on a site with a payment processor and pay later and the pay later option was broken.

**Comment by mattwire** [comment]
@monishdeb Looks like this needs a bit of work?

File: templates/CRM/common/paymentBlock.tpl (hunk 2 above)
Comment (monishdeb) [reply templates/CRM/common/paymentBlock.tpl:99]: @eileenmcnaughton yes I am passing the currency value for the 2nd case, if you look  at L122 below,  I am calling the buildPaymentBlock() fn on currency change and eventually at L99 we are considering the currency value and passing through the url  at L144

File: templates/CRM/common/paymentBlock.tpl (hunk 3 above)
Comment (monishdeb) [reply templates/CRM/common/paymentBlock.tpl]: Thanks @samuelsov for raising this issue and suggesting the fix. I have updated the PR with this commit https://github.com/civicrm/civicrm-core/pull/18517/commits/b86dc53eb2c51246435fedcc96b2ba947deea6c7 as per which we need to consider both the w/o pay-later cases.

**Comment by KarinG** [comment]
"Submit Credit Card Contribution backoffice form" -> it's not at all obvious that this is where one would go to process an ACH EFT (i.e. a non Credit Card) transaction.

**Comment by JoeMurray** [comment]
We were thinking of a separate PR for changing the name since it really should be Submit Online Contribution or something similar, @KarinG . What do you think?

**Comment by eileenmcnaughton** [comment]
I think you are brave :-) Changing button names is fighting talk. But yes, separate PR or issue don't mix it in with this

**Comment by KarinG** [comment]
I think this PR requires a new button -> "Process ACH EFT"

**Comment by KarinG** [comment]
@adixon - ping!

**Comment by eileenmcnaughton** [comment]
@KarinG no - a SEPARATE PR can cover that....

File: templates/CRM/common/paymentBlock.tpl [hunk 4]
```diff
@@ -111,15 +114,16 @@
...
       // Processors like pp-express will hide the form submit buttons, so re-show them when switching
       $('.crm-submit-buttons', $form).show().find('input').prop('disabled', true);
       CRM.loadPage(dataUrl, {target: '#billing-payment-block'});
     }
 
-    $('[name=payment_processor_id]').on('change.paymentBlock', function() {
-        buildPaymentBlock($(this).val());
+    $('[name=payment_processor_id], #currency').on('change.paymentBlock', function() {
+      var payment_processor_id = $('[name=payment_processor_id]:checked').val() == undefined ? $('[name=payment_processor_id]').val() : $('[name=payment_processor_id]:checked').val();
+      buildPaymentBlock(payment_processor_id);
```
Comment (mlutfy) [inline templates/CRM/common/paymentBlock.tpl:126]: @monishdeb This breaks backend New Contribution forms without a payment processor (ex: Contact Record -> New Contribution, then change the currency).

Adding "if (payment_contribution_id) { .. }" fixes it, although I haven't tested extensively.

For reference, this form:

civi-2020-11-06_13-44

**Comment by mattwire** [comment]
Hi @monishdeb This is awaiting changes following @mlutfy testing.

**Comment by monishdeb** [comment]
sorry for delay I have updated the PR.

**Comment by mlutfy** [comment]
Seems good to me, but I will let someone merge since it's not my area of expertise.

Symbiotic has been using this patch in production (with the latest commit) for a few weeks and seems to be working well.

**Comment by eileenmcnaughton** [comment]
I don't have any remaining concerns - @mattwire ?

**Comment by monishdeb** [comment]
ping @eileenmcnaughton @mattwire

**Pull request merged**